### PR TITLE
HIVE-28347: Make a UDAF 'collect_set' work with complex types, even when map-side aggregation is disabled.

### PR DESCRIPTION
**Pull request by JeongDaeKim**
### What changes were proposed in this pull request?

UDAFs, collect_set and collect_list, won't work with complex types, if map-side aggregation is  disabled. This PR will fix this bug.

### Why are the changes needed?

When users disable the map-side aggregation feature(`"hive.map.aggr`), collect_set doesn't work with complex types, such as struct.

### Does this PR introduce _any_ user-facing change?

No.

### Is the change a dependency upgrade?

No.

### How was this patch tested?

Tested with qtest (TestMiniTezCliDriver)

### CODE REVIEW

File: ql/src/java/org/apache/hadoop/hive/ql/udf/generic/GenericUDAFMkCollectionEvaluator.java
Comment (okumin) [inline ql/src/java/org/apache/hadoop/hive/ql/udf/generic/GenericUDAFMkCollectionEvaluator.java:79]: Can we merge this case with PARTIAL1?

```
if (m == Mode.PARTIAL1 || m == Mode.COMPLETE) {
  ...
```

**Comment by okumin** [comment]
@JeongDaeKim Taking a glance, your finding and fix seems to be very reasonable. Could you please regenerate the result file of `udaf_collect_set_2.q`?

**Comment by JeongDaeKim** [comment]
@okumin Thank you for a quick review! 👍

**Review by okumin — APPROVED** [review]
I have done three things.
First, I checked that the evaluator is likely correctly implemented now.
Second, I checked that we have new regrettion tests. Nice work.
Third, I ran the original `udaf_collect_set_2.q` with `set hive.map.aggr = true or false`. Both generated the same result set.

Then, the changes here look to me. Note that you also need +1 from a committer.

File: ql/src/java/org/apache/hadoop/hive/ql/udf/generic/GenericUDAFMkCollectionEvaluator.java
Comment (deniskuzZ) [inline ql/src/java/org/apache/hadoop/hive/ql/udf/generic/GenericUDAFMkCollectionEvaluator.java:70]: is it safe to remove
`!(parameters[0] instanceof ListObjectInspector)`
condition?

Comment (JeongDaeKim) [reply ql/src/java/org/apache/hadoop/hive/ql/udf/generic/GenericUDAFMkCollectionEvaluator.java:70]: Yes, I think that condition was just to check "no map aggregation" or not, not for safe type casting.
As comments said above, "The output of a partial aggregation is a list"

https://github.com/apache/hive/blob/588c2ad9bd343b0ec1e6fa761bc36c0e00682ee6/ql/src/java/org/apache/hadoop/hive/ql/udf/generic/GenericUDAFMkCollectionEvaluator.java#L43-L45

Comment (okumin) [reply ql/src/java/org/apache/hadoop/hive/ql/udf/generic/GenericUDAFMkCollectionEvaluator.java:70]: I also guess so. `COLLECT_LIST|SET` processes objects of a type T like that.
- `Mode.PARTIAL1` or `Mode.COMPLETE`: T => List[T]
- `Mode.PARTIAL2` or `Mode.FINAL`: List[T] => List[T]

The original one tried to check the type of inputs, T or List[T], to conclude the current mode. But it can lead to a wrong conclusion when T is List[X] because both `List[X]` and `List[List[X]]` hit the condition.

Comment (deniskuzZ) [reply ql/src/java/org/apache/hadoop/hive/ql/udf/generic/GenericUDAFMkCollectionEvaluator.java:70]: 👍, thanks for the explanation!

**Review by deniskuzZ — APPROVED** [review]
LGTM +1

File: ql/src/java/org/apache/hadoop/hive/ql/udf/generic/GenericUDAFMkCollectionEvaluator.java
Comment (deniskuzZ) [reply ql/src/java/org/apache/hadoop/hive/ql/udf/generic/GenericUDAFMkCollectionEvaluator.java:70]: could we update the comment with what @okumin wrote:
````
Mode.PARTIAL1 or Mode.COMPLETE: T => List[T]
Mode.PARTIAL2 or Mode.FINAL: List[T] => List[T]
````
`The output of a partial aggregation is a list` is kinda confusing, PARTIAL2 is also partial

see comments at the top
// For PARTIAL1 and COMPLETE: ObjectInspectors for original data
// For PARTIAL2 and FINAL: ObjectInspectors for partial aggregations (list of objs)

Comment (JeongDaeKim) [reply ql/src/java/org/apache/hadoop/hive/ql/udf/generic/GenericUDAFMkCollectionEvaluator.java:72]: https://github.com/apache/hive/pull/5323/commits/74e7ce7087682c2a9c23a25b9c1423840baf3b0a

Comments updated. (I don't know this comments are necessary though.)

Comment (deniskuzZ) [reply ql/src/java/org/apache/hadoop/hive/ql/udf/generic/GenericUDAFMkCollectionEvaluator.java:72]: @JeongDaeKim , sorry fo the confusion, no need for `// The output of a partial aggregation is a list, output of COMPLETE is List as well`. Just keep the mapping explanation

File: ql/src/java/org/apache/hadoop/hive/ql/udf/generic/GenericUDAFMkCollectionEvaluator.java
Comment (deniskuzZ) [inline ql/src/java/org/apache/hadoop/hive/ql/udf/generic/GenericUDAFMkCollectionEvaluator.java:66]: could we please inline the comments
````
if (mode == Mode.PARTIAL1 || mode == Mode.COMPLETE) {
      // T => List[T]
      inputOI = parameters[0];
      return ObjectInspectorFactory.getStandardListObjectInspector(
          ObjectInspectorUtils.getStandardObjectInspector(inputOI));
    } else {
      // List[T] => List[T]
      ....
}
````

**Comment by deniskuzZ** [comment]
thanks @JeongDaeKim! I'll merge it with the next green build

**Pull request merged**
